### PR TITLE
circle invoice payment split

### PR DESCRIPTION
**Pull request by Carolusian**

### CODE REVIEW

File: db/migrations/20210119152027_create_circle_invoice_table.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+const { baseDown } = require('../utils')
+
+const table = 'circle_invoice'
+
+exports.up = async (knex) => {
+  await knex('entity_type').insert({ table })
+  await knex.schema.createTable(table, (t) => {
+    t.bigIncrements('id').primary()
+    t.bigInteger('user_id').unsigned().notNullable
```
Comment (robertu7) [inline db/migrations/20210119152027_create_circle_invoice_table.js:9]: `.notNullable()`?

File: src/connectors/paymentService.ts [hunk 2]
```diff
@@ -490,4 +492,143 @@ export class PaymentService extends BaseService {
...
+          provider: PAYMENT_PROVIDER.stripe,
+          providerInvoiceId,
+        })
+        .returning('id')
+
+      // verify if it is OK to split the payment given the circle prices
+      for (const p of prices) {
+        if (p.currency !== currency) {
+          throw new ServerError(
+            `currency for '${providerTxId}' if not the same to '${p.providerPriceId}'`
```
Comment (robertu7) [inline src/connectors/paymentService.ts:583]: typo? `if` -> `is`?

File: src/routes/pay/stripe.ts [hunk 3]
```diff
@@ -107,6 +113,65 @@ const updateTxState = async (
...
+    const providerTxId = invoice.payment_intent as string
+    const amount = toDBAmount({ amount: invoice.amount_paid })
+    const currency = _.toUpper(invoice.currency) as PAYMENT_CURRENCY
+    const tx = (await paymentService.findTransactions({ providerTxId }))[0]
+    // retrieve customer
+    const customer = (await atomService.findFirst({
+      table: 'customer',
+      where: {
+        customer_id: invoice.customer,
+        provider: PAYMENT_PROVIDER.stripe,
```
Comment (robertu7) [inline src/routes/pay/stripe.ts:137]: missing `archived: false`?

File: src/routes/pay/stripe.ts [hunk 4]
```diff
@@ -107,6 +113,65 @@ const updateTxState = async (
...
+) => {
+  logger.info(`proceeding ${eventType} event...`)
+  if (eventType === 'invoice.payment_succeeded') {
+    const atomService = new AtomService()
+    const paymentService = new PaymentService()
+    const providerInvoiceId = invoice.id as string
+    const providerTxId = invoice.payment_intent as string
+    const amount = toDBAmount({ amount: invoice.amount_paid })
+    const currency = _.toUpper(invoice.currency) as PAYMENT_CURRENCY
+    const tx = (await paymentService.findTransactions({ providerTxId }))[0]
```
Comment (robertu7) [inline src/routes/pay/stripe.ts:131]: Same as below, save some DB costs if return first (`if (!tx) { return }`)?

**Review by robertu7 — CHANGES_REQUESTED** [review]
Left some comments, others look good to me.

File: src/connectors/paymentService.ts [hunk 5]
```diff
@@ -490,4 +492,143 @@ export class PaymentService extends BaseService {
...
+          `sum of plan prices '${totalAmount}' != invoice paid amount '${amount}'`
+        )
+      }
+
+      // create transactions to split the invoice payment to circles
+      const { id: entityTypeId } = await this.baseFindEntityTypeId(
+        TRANSACTION_TARGET_TYPE.circlePrice
+      )
+      for (const p of prices) {
+        const circle = await this.atomService.findFirst({
```
Comment (devformatters) [inline src/connectors/paymentService.ts:602]: I would not recommend to init atomService in another service. You can get rid of `atomService` by using:
- `this.baseFindById`  (paymentService extends baseService)
- `this.knex().select().where().first()`

**Review by devformatters — CHANGES_REQUESTED** [review]
Left one comment.

**Review by guoliu — APPROVED** [review]
Looks good, let's try it 🍀

File: src/connectors/paymentService.ts [hunk 6]
```diff
@@ -491,6 +492,140 @@ export class PaymentService extends BaseService {
...
+          `sum of plan prices '${totalAmount}' != invoice paid amount '${amount}'`
+        )
+      }
+
+      // create transactions to split the invoice payment to circles
+      const { id: entityTypeId } = await this.baseFindEntityTypeId(
+        TRANSACTION_TARGET_TYPE.circlePrice
+      )
+      for (const p of prices) {
+        const circle = await this.baseFindById(p.circleId.toString(), 'circle')
```
Comment (robertu7) [inline src/connectors/paymentService.ts:601]: `.toString()` is unnecessary, i beleve.

**Review by robertu7 — APPROVED** [review]

File: db/migrations/20210119152027_create_circle_invoice_table.js (hunk 1 above)
Comment (Carolusian) [reply db/migrations/20210119152027_create_circle_invoice_table.js]: updated.

**Review by devformatters — APPROVED** [review]
Go. 🦊

**Pull request merged**